tket/src/Transformations/include/Transformations/PhasedXFrontier.hpp

```diff
@@ -0,0 +1,290 @@
+// Copyright 2019-2022 Cambridge Quantum Computing
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cmath>
+#include <set>
+
+#include "Circuit/Circuit.hpp"
+#include "Ops/OpPtr.hpp"
+#include "SingleQubitSquash.hpp"
+
+namespace tket {
+
+namespace Transforms {
+
+// type aliases
+using OptVertex = std::optional<Vertex>;
+using OptVertexVec = std::vector<OptVertex>;
+using OptEdge = std::optional<Edge>;
+using OptEdgeVec = std::vector<OptEdge>;
+
+/**
+ * @brief Frontier and circuit manipulation for `globalise_PhasedX`
+ *
+ * This helper class for the \ref globalise_PhasedX transform stores the circuit
+ * frontier as useful for that specific pass. It also provides a simple
+ * interface for all the circuit manipulations that are required within that
+ * pass.
+ *
+ * The transform itself therefore only needs to worry about the logic of the
+ * transformation and none of the bookkeeping and circuit substitutions.
+ *
+ * All gates of `circ` must be either PhasedX, NPhasedX, Rz or multi-qubit
+ * gates.
+ *
+ * ## Intervals
+ * For each qubit, the frontier stores the current interval. An interval is a
+ * single-qubit subcircuit of `circ` forming a sequence of single-qubit gates
+ * between two multi-qubit gates (or non-gate vertices) of the original
+ * circuit. Each interval is defined by start and end edges, which are edges of
+ * the circuit DAG ("end" is in the future of "start"). The source of the start
+ * edge is a multi-qubit gate (or a non-gate vertex such as OpType::Input or
+ * OpType::Create) and the target of the end
```

Comment:
```suggestion
 * edge is a multi-qubit gate (or an initial vertex) and the target of the end
```

Reply:
I'm not sure I understand this suggestion. Were you trying to remove some redundancy in the sentence? I've improved the phrasing a bit.

Reply:
OK I understand now, thanks.